### PR TITLE
#103 Change some of the members in red_black_tree from public to protected

### PR DESCRIPTION
**Pull request by HenryRLee**
This commit is a solution to issue #103 
These changes include:
1. Change some members in red_black_tree from public to protected
2. Add friend declaration in class rbtree_iterator so that rbtree can access its private member
3. Some of the lines in map and set were using the iterator incorrectly. Instead of accessing mpNode and mValue, use the -> and * operators in the iterator directly.

### CODE REVIEW

File: include/EASTL/internal/red_black_tree.h [hunk 1]
```diff
@@ -184,6 +184,22 @@ namespace eastl
 		rbtree_iterator& operator--();
 		rbtree_iterator  operator--(int);
 
+	protected:
+		template <typename Tp, typename PointerA, typename ReferenceA, typename PointerB, typename ReferenceB>
+		friend bool operator==(const rbtree_iterator<Tp, PointerA, ReferenceA>& a,
```
Comment (rparolin) [inline include/EASTL/internal/red_black_tree.h]: Where are the implementations for these operators?

Comment (rparolin) [inline include/EASTL/internal/red_black_tree.h]: Also, I'd rather use the class member function variant vs friending a freestanding functions.

Comment (rparolin) [inline include/EASTL/internal/red_black_tree.h]: Also, why do you need to add these comparison operators?  Changing access control doesn't appear to warrant it.

**Comment by rparolin** [comment]
Other than the comments about the friend function.   I think the changes look good on their own.  I'm not sure how much client code I will break by limiting access to these variables.

File: include/EASTL/internal/red_black_tree.h (hunk 1 above)
Comment (HenryRLee) [inline include/EASTL/internal/red_black_tree.h]: Hi, because these global functions already existed, between line 677 and 701. I just tried to make minimum changes to the original code.

Comment (HenryRLee) [inline include/EASTL/internal/red_black_tree.h]: So basically they are friend forward declarations of these functions. https://github.com/electronicarts/EASTL/blob/master/include/EASTL/internal/red_black_tree.h#L675

Comment (rparolin) [inline include/EASTL/internal/red_black_tree.h]: Why did they need to be added here?

Comment (HenryRLee) [inline include/EASTL/internal/red_black_tree.h]: You mean why the original author wrote these functions? Because they wanted to have comparison between `iterator` and `const_iterator`.

Comment (rparolin) [inline include/EASTL/internal/red_black_tree.h]: No, I mean why did you add the friend declarations for this change?

Comment (HenryRLee) [inline include/EASTL/internal/red_black_tree.h]: Because these functions are accessing the internal node pointer, which is declared as protected now.

**Comment by rparolin** [comment]
Apologises for not making this decision sooner (as I've gone back and forth on it) but I'm going to keep the code as-is for simplicity and legacy considerations.  Thanks for the work and apologises again.

**Comment by HenryRLee** [comment]
I see. No problem.
